rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,198 @@
+import java.util.concurrent.ConcurrentMap;
+
+@Slf4j
+@RuleNode(
+        type = ComponentType.ACTION,
+        name = "device state",
+        nodeDescription = "Triggers device connectivity events",
+        nodeDetails = "If incoming message originator is a device, registers configured event for that device in the Device State Service, which sends appropriate message to the Rule Engine." +
+                " If metadata <code>ts</code> property is present, it will be used as event timestamp. Otherwise, the message timestamp will be used." +
+                " Incoming message is forwarded using the <code>Success</code> chain, unless an unexpected error occurs during message processing" +
+                " then incoming message is forwarded using the <code>Failure</code> chain." +
+                "<br>" +
+                "Supported device connectivity events are:" +
+                "<ul>" +
+                "<li>Connect event</li>" +
+                "<li>Disconnect event</li>" +
+                "<li>Activity event</li>" +
+                "<li>Inactivity event</li>" +
+                "</ul>" +
+                "This node is particularly useful when device isn't using transports to receive data, such as when fetching data from external API or computing new data within the rule chain.",
+        configClazz = TbDeviceStateNodeConfiguration.class,
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbActionNodeDeviceStateConfig"
+)
+public class TbDeviceStateNode implements TbNode {
+
+    private static final Set<TbMsgType> SUPPORTED_EVENTS = EnumSet.of(
+            TbMsgType.CONNECT_EVENT, TbMsgType.ACTIVITY_EVENT, TbMsgType.DISCONNECT_EVENT, TbMsgType.INACTIVITY_EVENT
+    );
+    private static final Duration ONE_SECOND = Duration.ofSeconds(1L);
+    private static final Duration ENTRY_EXPIRATION_TIME = Duration.ofDays(1L);
+
+    private Stopwatch stopwatch;
+    private ConcurrentMap<DeviceId, Duration> lastActivityEventTimestamps;
+    private TbMsgType event;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        TbMsgType event = TbNodeUtils.convert(configuration, TbDeviceStateNodeConfiguration.class).getEvent();
+        if (event == null) {
+            throw new TbNodeException("Event cannot be null!", true);
+        }
+        if (!SUPPORTED_EVENTS.contains(event)) {
+            throw new TbNodeException("Unsupported event: " + event, true);
+        }
+        this.event = event;
+        lastActivityEventTimestamps = new ConcurrentHashMap<>();
+        stopwatch = Stopwatch.createStarted();
+        scheduleCleanupMsg(ctx);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) {
+        if (msg.isTypeOf(TbMsgType.DEVICE_STATE_STALE_ENTRIES_CLEANUP_MSG)) {
+            removeStaleEntries();
+            scheduleCleanupMsg(ctx);
+            ctx.ack(msg);
```

Comment:
this message is not persisted in the queue so no needs to ack it.

Reply:
Fixed in https://github.com/thingsboard/thingsboard/pull/9030/commits/691a6b5d6a1cc0acd300eca5d9396e296207bf82